File: internal/license/verify.go
package license

import (
	"crypto/ed25519"
	"errors"
	"fmt"
	"os"
	"time"
)

// VerifyOptions wires the verifier to its on-disk inputs.
type VerifyOptions struct {
	LicensePath   string
	PublicKey     ed25519.PublicKey
	WatermarkPath string
	Fingerprint   FingerprintConfig

	// Now lets tests inject the wall clock. Production callers leave
	// this zero so time.Now().UTC() is used.
	Now time.Time

	// Logger is invoked once with a structured event describing the
	// verification outcome. May be nil.
	Logger func(event VerifyEvent)
}

// VerifyEvent is what the optional Logger receives. We deliberately
// keep the details out of VerifyResult so the public Status API can
// stay opaque while operators still get useful diagnostics on the box.
type VerifyEvent struct {
	Reason      VerifyReason
	Err         error
	LicenseID   string
	Fingerprint string
	Now         time.Time
	Effective   time.Time
}

// Verify is the single source of truth. The Next.js side calls into
// either this function (when embedding Go via cgo) or its TypeScript
// twin in verifier-node.
func Verify(opts VerifyOptions) VerifyResult {
	now := opts.Now
	if now.IsZero() {
		now = time.Now().UTC()
	}
	log := opts.Logger
	emit := func(reason VerifyReason, err error, licID, fp string, eff time.Time) VerifyResult {
		if log != nil {
			log(VerifyEvent{Reason: reason, Err: err, LicenseID: licID, Fingerprint: fp, Now: now, Effective: eff})
		}
		res := VerifyResult{Valid: reason == ReasonOK, Reason: reason}
		if reason == ReasonOK {
			res.LicenseID = licID
		}
		return res
	}

	// 1. Read and structurally parse the license.
	var lic License
	if err := ReadJSONFileStrict(opts.LicensePath, &lic); err != nil {
		return emit(ReasonMalformed, err, "", "", time.Time{})
	}
	if lic.Version != FormatVersion {
		return emit(ReasonUnsupportedVersion,
			fmt.Errorf("got version %d, want %d", lic.Version, FormatVersion), lic.ID, "", time.Time{})
	}
	if lic.ID == "" || lic.HardwareFingerprint == "" || lic.NotAfter.IsZero() || lic.IssuedAt.IsZero() {
		return emit(ReasonMalformed, errors.New("missing required field"), lic.ID, "", time.Time{})
	}

	// 2. Signature check first — it's the cheapest hard gate.
	if err := VerifySignature(opts.PublicKey, &lic); err != nil {
		return emit(ReasonSignatureInvalid, err, lic.ID, "", time.Time{})
	}

	// 3. Hardware fingerprint.
	info, err := CollectHardwareInfo(opts.Fingerprint)
	if err != nil {
		return emit(ReasonHardwareUnavailable, err, lic.ID, "", time.Time{})
	}
	if info.Fingerprint != lic.HardwareFingerprint {
		err := fmt.Errorf("device fp %s != license fp %s", info.Fingerprint, lic.HardwareFingerprint)
		if os.Geteuid() != 0 {
			if locked := RootLockedDMISourceKeys(opts.Fingerprint); len(locked) > 0 {
				err = fmt.Errorf("%w; unreadable root-only DMI fields %v (retry with sudo)", err, locked)
			}
		}
		return emit(ReasonFingerprintMismatch, err, lic.ID, info.Fingerprint, time.Time{})
	}

	// 4. Derive keys.
	fpBytes, err := FingerprintBytes(lic.HardwareFingerprint)
	if err != nil {
		return emit(ReasonMalformed, err, lic.ID, info.Fingerprint, time.Time{})
	}
	aesKey, err := DeriveKey(fpBytes, lic.ID, HKDFInfoPayloadKey)
	if err != nil {
		return emit(ReasonMalformed, err, lic.ID, info.Fingerprint, time.Time{})
	}
	defer zero(aesKey)
	hmacKey, err := DeriveKey(fpBytes, lic.ID, HKDFInfoWatermarkKey)
	if err != nil {
		return emit(ReasonMalformed, err, lic.ID, info.Fingerprint, time.Time{})
	}
	defer zero(hmacKey)

	// 5. Decrypt + cross-bind payload to header.
	payload, err := DecryptPayload(aesKey, lic.EncryptedPayload)
	if err != nil {
		return emit(ReasonMalformed, err, lic.ID, info.Fingerprint, time.Time{})
	}
	if payload.ID != lic.ID {
		return emit(ReasonPayloadMismatch,
			fmt.Errorf("payload.id %s != license.id %s", payload.ID, lic.ID),
			lic.ID, info.Fingerprint, time.Time{})
	}
	if !payload.NotAfter.Equal(lic.NotAfter) {
		return emit(ReasonPayloadMismatch,
			fmt.Errorf("payload.notAfter != license.notAfter"),
			lic.ID, info.Fingerprint, time.Time{})
	}

	// 6. Watermark.
	wm, err := LoadWatermark(opts.WatermarkPath, hmacKey)
	if err != nil {
		return emit(ReasonWatermarkTampered, err, lic.ID, info.Fingerprint, time.Time{})
	}
	if wm != nil {
		if !wm.SameLicense(&lic) {
			return emit(ReasonWatermarkTampered,
				fmt.Errorf("watermark belongs to license %s, current is %s", wm.LicenseID, lic.ID),
				lic.ID, info.Fingerprint, time.Time{})
		}
		if err := wm.MustValidate(); err != nil {
			return emit(ReasonWatermarkTampered, err, lic.ID, info.Fingerprint, time.Time{})
		}
	}

	next, effective, _ := AdvanceWatermark(wm, &lic, now)

	if next.TimeRewindCount > MaxTimeRewinds {
		// We still persist the bumped counter so an attacker can't make
		// progress by repeatedly killing the process before the file
		// gets re-written.
		_ = SaveWatermark(opts.WatermarkPath, hmacKey, next)
		return emit(ReasonTimeRewind,
			fmt.Errorf("time_rewind_count=%d > %d", next.TimeRewindCount, MaxTimeRewinds),
			lic.ID, info.Fingerprint, effective)
	}

	// 7. Time gates against the *effective* clock.
	if effective.Before(lic.NotBefore) {
		_ = SaveWatermark(opts.WatermarkPath, hmacKey, next)
		return emit(ReasonNotYetValid, nil, lic.ID, info.Fingerprint, effective)
	}
	if effective.After(lic.NotAfter) {
		_ = SaveWatermark(opts.WatermarkPath, hmacKey, next)
		return emit(ReasonExpired, nil, lic.ID, info.Fingerprint, effective)
	}

	// 8. Offline-duration cap (compares against the *real* wall clock,
	//    not the effective clock — we only want to know "how long since
	//    we last saw a fresh license"; that requires real-time deltas).
	if payload.MaxOfflineDays > 0 && wm != nil {
		gap := now.Sub(wm.LastSeenAt)
		if gap.Hours()/24 > float64(payload.MaxOfflineDays) {
			_ = SaveWatermark(opts.WatermarkPath, hmacKey, next)
			return emit(ReasonOfflineTooLong,
				fmt.Errorf("gap=%s > %dd", gap, payload.MaxOfflineDays),
				lic.ID, info.Fingerprint, effective)
		}
	}

	// 9. Success — persist updated watermark, build the public result.
	if err := SaveWatermark(opts.WatermarkPath, hmacKey, next); err != nil {
		return emit(ReasonMalformed, fmt.Errorf("save watermark: %w", err),
			lic.ID, info.Fingerprint, effective)
	}

	days := int(lic.NotAfter.Sub(effective).Hours() / 24)
	notAfter := lic.NotAfter
	if log != nil {
		log(VerifyEvent{Reason: ReasonOK, LicenseID: lic.ID, Fingerprint: info.Fingerprint, Now: now, Effective: effective})
	}
	return VerifyResult{
		Valid:     true,
		Reason:    ReasonOK,
		NotAfter:  &notAfter,
		DaysLeft:  &days,
		LicenseID: lic.ID,
		Features:  payload.Features,
	}
}

File: verifier-node/src/verify.ts
// Public verification entrypoint. Mirrors internal/license/verify.go.
//
// Verify() is the only function Next.js should call. It is fully
// synchronous: there is no I/O fan-out, no network, no spawn other
// than nvidia-smi (which is part of fingerprinting). Verification of a
// healthy license on commodity hardware takes ~3 ms.

import { readFileSync } from 'node:fs';

import {
  License,
  VerifyResult,
  VerifyReason,
  FORMAT_VERSION,
  HKDF_INFO_PAYLOAD,
  HKDF_INFO_WATERMARK,
  MAX_TIME_REWINDS,
} from './types';
import {
  decryptPayload,
  deriveKey,
  fingerprintBytes,
  loadEd25519Public,
  verifyLicenseSignature,
} from './crypto';
import {
  FingerprintConfig,
  collectHardwareInfo,
  fingerprintConfigFromEnv,
} from './fingerprint';
import {
  advanceWatermark,
  loadWatermark,
  saveWatermark,
  validateWatermark,
  watermarkBelongsTo,
} from './watermark';

export interface VerifyOptions {
  licensePath: string;
  publicKeyPath: string;
  watermarkPath: string;
  fingerprint?: Partial<FingerprintConfig>;
  /** Override the wall clock; production callers leave undefined. */
  now?: Date;
  /** Optional diagnostic sink; the public VerifyResult deliberately omits details. */
  logger?: (event: VerifyEvent) => void;
}

export interface VerifyEvent {
  reason: VerifyReason;
  error?: Error;
  licenseId?: string;
  fingerprint?: string;
  now: Date;
  effective?: Date;
}

export function verifyLicense(opts: VerifyOptions): VerifyResult {
  const now = opts.now ?? new Date();
  const logger = opts.logger ?? (() => {});
  // Bare-metal host paths by default; switch to container bind-mount
  // paths only when HW_CONTAINER=1 (mirrors the Go CLI). Caller-provided
  // `fingerprint` overrides win regardless.
  const cfg: FingerprintConfig = { ...fingerprintConfigFromEnv(), ...(opts.fingerprint ?? {}) };

  const emit = (reason: VerifyReason, error?: Error, licenseId?: string, fingerprint?: string, effective?: Date): VerifyResult => {
    logger({ reason, error, licenseId, fingerprint, now, effective });
    return {
      valid: reason === 'ok',
      reason,
      licenseId: reason === 'ok' ? licenseId : undefined,
    };
  };

  // 1. Read + structurally validate the license.
  let lic: License;
  try {
    const raw = readFileSync(opts.licensePath, 'utf8');
    lic = JSON.parse(raw) as License;
  } catch (e) {
    return emit('malformed', e as Error);
  }
  if (lic.version !== FORMAT_VERSION) {
    return emit('unsupported_version', new Error(`got ${lic.version}, want ${FORMAT_VERSION}`), lic.id);
  }
  if (!lic.id || !lic.hardwareFingerprint || !lic.issuedAt) {
    return emit('malformed', new Error('missing required field'), lic.id);
  }
  // Permanent licenses carry the Go zero time ("0001-01-01T00:00:00Z");
  // only require a real notAfter for time-bound ones.
  if (lic.expires && !lic.notAfter) {
    return emit('malformed', new Error('expiring license missing notAfter'), lic.id);
  }

  // 2. Signature check.
  let pub;
  try {
    pub = loadEd25519Public(opts.publicKeyPath);
  } catch (e) {
    return emit('malformed', e as Error, lic.id);
  }
  try {
    verifyLicenseSignature(pub, lic);
  } catch (e) {
    return emit('signature_invalid', e as Error, lic.id);
  }

  // 3. Hardware fingerprint.
  let info;
  try {
    info = collectHardwareInfo(cfg);
  } catch (e) {
    return emit('hardware_unavailable', e as Error, lic.id);
  }
  if (info.fingerprint !== lic.hardwareFingerprint) {
    return emit('fingerprint_mismatch',
      new Error(`device fp ${info.fingerprint} != license fp ${lic.hardwareFingerprint}`),
      lic.id, info.fingerprint);
  }

  // 4. Derive keys.
  const fpBuf = fingerprintBytes(lic.hardwareFingerprint);
  const aesKey = deriveKey(fpBuf, lic.id, HKDF_INFO_PAYLOAD);
  const hmacKey = deriveKey(fpBuf, lic.id, HKDF_INFO_WATERMARK);

  // 5. Decrypt + payload cross-binding.
  let payload;
  try {
    payload = decryptPayload(aesKey, lic.encryptedPayload);
  } catch (e) {
    return emit('malformed', e as Error, lic.id, info.fingerprint);
  }
  if (payload.id !== lic.id) {
    return emit('payload_mismatch',
      new Error(`payload.id=${payload.id} != license.id=${lic.id}`),
      lic.id, info.fingerprint);
  }
  if (payload.expires !== lic.expires) {
    return emit('payload_mismatch',
      new Error(`payload.expires=${payload.expires} != license.expires=${lic.expires}`),
      lic.id, info.fingerprint);
  }
  if (payload.notAfter !== lic.notAfter) {
    return emit('payload_mismatch',
      new Error('payload.notAfter != license.notAfter'),
      lic.id, info.fingerprint);
  }

  // 6. Watermark.
  let wm;
  try {
    wm = loadWatermark(opts.watermarkPath, hmacKey);
  } catch (e) {
    return emit('watermark_tampered', e as Error, lic.id, info.fingerprint);
  }
  if (wm) {
    if (!watermarkBelongsTo(wm, lic)) {
      return emit('watermark_tampered',
        new Error(`watermark belongs to ${wm.licenseId}, not ${lic.id}`),
        lic.id, info.fingerprint);
    }
    try {
      validateWatermark(wm);
    } catch (e) {
      return emit('watermark_tampered', e as Error, lic.id, info.fingerprint);
    }
  }

  const adv = advanceWatermark(wm, lic, now);

  if (adv.next.timeRewindCount > MAX_TIME_REWINDS) {
    safeSave(opts.watermarkPath, hmacKey, adv.next);
    return emit('time_rewind',
      new Error(`time_rewind_count=${adv.next.timeRewindCount} > ${MAX_TIME_REWINDS}`),
      lic.id, info.fingerprint, adv.effective);
  }

  // 7. Time gates against the effective clock. NotBefore is enforced
  //    even for permanent licenses; the expired check only runs when
  //    lic.expires is true.
  const notBefore = new Date(lic.notBefore);
  if (adv.effective.getTime() < notBefore.getTime()) {
    safeSave(opts.watermarkPath, hmacKey, adv.next);
    return emit('not_yet_valid', undefined, lic.id, info.fingerprint, adv.effective);
  }
  const notAfter = lic.expires ? new Date(lic.notAfter) : null;
  if (notAfter && adv.effective.getTime() > notAfter.getTime()) {
    safeSave(opts.watermarkPath, hmacKey, adv.next);
    return emit('expired', undefined, lic.id, info.fingerprint, adv.effective);
  }

  // 8. Offline-duration cap, against real wall clock. Skipped for
  //    permanent licenses (the issuer forces maxOfflineDays=0 in that
  //    mode, but we guard explicitly in case a hand-crafted payload
  //    slips through).
  if (lic.expires && payload.maxOfflineDays > 0 && wm) {
    const gapMs = now.getTime() - new Date(wm.lastSeenAt).getTime();
    const gapDays = gapMs / 86_400_000;
    if (gapDays > payload.maxOfflineDays) {
      safeSave(opts.watermarkPath, hmacKey, adv.next);
      return emit('offline_too_long',
        new Error(`gap=${gapDays.toFixed(2)}d > ${payload.maxOfflineDays}d`),
        lic.id, info.fingerprint, adv.effective);
    }
  }

  // 9. Persist + return.
  try {
    saveWatermark(opts.watermarkPath, hmacKey, adv.next);
  } catch (e) {
    return emit('watermark_tampered', e as Error, lic.id, info.fingerprint, adv.effective);
  }

  logger({ reason: 'ok', licenseId: lic.id, fingerprint: info.fingerprint, now, effective: adv.effective });
  const result: VerifyResult = {
    valid: true,
    reason: 'ok',
    licenseId: lic.id,
    features: payload.features,
  };
  if (notAfter) {
    result.notAfter = lic.notAfter;
    result.daysLeft = Math.floor((notAfter.getTime() - adv.effective.getTime()) / 86_400_000);
  }
  return result;
}

function safeSave(filePath: string, key: Buffer, wm: import('./types').Watermark): void {
  try {
    saveWatermark(filePath, key, wm);
  } catch {
    // swallow — we already failed verification, this is best-effort persistence
  }
}
